### ISSUE TITLE
Validation loss unexpectedly low

### ISSUE THREAD

**Issue by iipr** [keras-team/keras#13591]
# System information
- Have I written custom code (as opposed to using example directory):  Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 18.04 
- TensorFlow backend (yes / no):  Yes
- TensorFlow version:  2.0.0 (git version: v2.0.0-rc2-26-g64c3d38)
- Keras version:  2.3.1
- Python version:  3.7.5
- CUDA/cuDNN version: CUDA 10.1 (not used) 
- GPU model and memory: - (not used)

# Describe the current behavior

When training a simple model, the validation loss seemed too low, orders of magnitude lower than the training loss. I wrote some additional code to keep track of the loss per batch, both at training and validation. I may be utterly wrong, but to me it seems like the validation loss displayed once the validation step is finished is not the average over batches, but rather the loss of the last validation batch. I will provide information about this intuition later on.

_Note:_ I have read in the [Keras FAQ](https://keras.io/getting-started/faq/#why-is-the-training-loss-much-higher-than-the-testing-loss) and some StackOverflow posts that this is due to regularization mechanisms, but I am not using any regularization method, or batch normalization, or dropout. The model I tested is a simple network with just dense layers.

# Describe the expected behavior

For every epoch and after completion of the validation step, show the average over the losses of the validation batches.  

# Code to reproduce the issue

## The model

```
import numpy as np
from keras import layers, models, optimizers

model = models.Sequential()
model.add(layers.Reshape((170,), input_shape=(17,10)))
model.add(layers.Dense(68, activation='relu'))
model.add(layers.Dense(136, activation='relu'))
model.add(layers.Dense(68, activation='relu'))
model.add(layers.Dense(17, activation='linear'))

model.compile(optimizer=optimizers.RMSprop(lr=1e-4), 
              loss='mse', metrics=['mae', 'mse'])
```

## Data generator

I use a data generator to produce the batches of data. Simplified version:

```
from keras.utils import Sequence

class DataGenerator(Sequence):
    
    def __init__(data, batch_size=2**6):
        self.data = data
        self.batch_size = batch_size

    def __len__(self):
        return self.data.shape[0] // self.batch_size

    def __getitem__(self, batch_n):
        return data[batch_n*self.batch_size:(batch_n + 1)*self.batch_size]
```

## History callback 

I use as well a history callback to save losses and metrics to analyse them afterwards. Simplified version:

```
import pandas as pd
from keras.callbacks import Callback

class HistoryCallback(Callback):
    
    def __init__(self):
        pass

    def on_epoch_begin(self, epoch, logs={}):
        # Define to tables to fill them after every batch
        self.batch_end = pd.DataFrame(index=range(10000))
        self.batch_end_val = pd.DataFrame(index=range(3000))
        
    def on_batch_end(self, batch, logs={}):
        for k in logs.keys():
            self.batch_end.loc[batch, k] = logs[k]

    def on_test_batch_end(self, batch, logs={}):
        for k in logs.keys():
            self.batch_end_val.loc[batch, k] = logs[k]
            
    def on_epoch_end(self, epoch, logs={}):
        self.batch_end.to_csv('./loss_per_batch_ep{}.csv'.format(epoch))
        self.batch_end_val.to_csv('./loss_per_batch_val_ep{}.csv'.format(epoch))
```

## Training

Something to mention here. The data I am working with is time dependent: it starts low at the beggining of the day, grows during the day and plummets towards the end of the day (that is why in the upcoming plots the error will have this behaviour). It is recorded for every day of several months, so for every month I take some days for training and some for validation.

```
for epoch in range(epochs):
    for month in range(months):
        n_days = days_in_month(month)
        train_days = np.arange(0, 25)
        val_days = np.arange(25, n_days)
        train_generator = DataGenerator(train_days, **params)
        val_generator = DataGenerator(val_days, **params)
        model.fit_generator(generator=train_generator,
                            validation_data=val_generator, shuffle=False,
                            initial_epoch=epoch, epochs=epoch+1,
                            callbacks=[HistoryCallback()],
                            use_multiprocessing=True, workers=0)
```

## Analysis

After completion of the training, we can now plot the loss and metrics. With 3 epochs and 17 months, there are 51 values to plot. Recall that `loss='mse', metrics=['mae', 'mse']`.

![image](https://user-images.githubusercontent.com/17690149/69868089-08861780-12a9-11ea-8c9c-4f84727962b1.png)

First question: **shouldn't the 2nd and 3rd plots be the same?**  
They both show the MSE, the 2nd one comes from metric and the 3rd one from loss.
As you can see, values are the same for training but greatly differ for validation.

That is why I added to the `HistoryCallback()` the lines to keep track of the loss per batch and save them into a CSV on epoch end. Analysing them: 

### Loss per batch for training

```
df = pd.read_csv('./loss_per_batch_ep1.csv', index_col=0)
display(df.head())
fig, ax = plt.subplots()
fig.set_size_inches(20, 10)
ax.plot(df[['loss', 'mae', 'mse']])
ax.set_xlabel('batch ID', fontsize=30)
ax.xaxis.set_tick_params(labelsize=20)
ax.set_ylabel('loss/metric', fontsize=30)
ax.yaxis.set_tick_params(labelsize=20)
ax.legend(['loss', 'mae', 'mse'], fontsize=25)
ax.grid()
ax.set_title('Loss (MSE) and metrics (MSE and MAE) for training', fontsize=30)
plt.show()
```
![image](https://user-images.githubusercontent.com/17690149/69871654-a848a300-12b3-11ea-94c1-9ff525f5691e.png)

![image](https://user-images.githubusercontent.com/17690149/69871639-9f57d180-12b3-11ea-9105-f154551d4b6f.png)

Checking the mean of loss:

```
print(df.loss.mean())
>>> 0.03531155811965683
```

Checking the tail of the loss and metrics per batch:

```
df[~df.loss.isna()][['loss', 'mae', 'mse']].tail()
 ```
![image](https://user-images.githubusercontent.com/17690149/69871791-01b0d200-12b4-11ea-8906-cb39db5ef5ec.png)

Note that the last value of the column `mse` is the same as the mean of the column `loss`.

### Loss per batch for validation
```
df = pd.read_csv('./loss_per_batch_val_ep1.csv', index_col=0)
display(df.head())
fig, ax = plt.subplots()
fig.set_size_inches(20, 10)
ax.plot(df[['loss', 'mae', 'mse']])
ax.set_xlabel('batch ID', fontsize=30)
ax.xaxis.set_tick_params(labelsize=20)
ax.set_ylabel('loss/metric', fontsize=30)
ax.yaxis.set_tick_params(labelsize=20)
ax.legend(['loss', 'mae', 'mse'], fontsize=25)
ax.grid()
ax.set_title('Loss (MSE) and metrics (MSE and MAE) for validation', fontsize=30)
plt.show()
```
![image](https://user-images.githubusercontent.com/17690149/69871845-26a54500-12b4-11ea-9a00-91c9a4421050.png)

![image](https://user-images.githubusercontent.com/17690149/69871849-2ad16280-12b4-11ea-917f-91df2ad1f81f.png)

Checking the mean of loss:

```
print(df.loss.mean())
>>> 0.051740424850634396
```

Checking the tail of the loss and metrics per batch:

```
df[~df.loss.isna()][['loss', 'mae', 'mse']].tail()
 ```
![image](https://user-images.githubusercontent.com/17690149/69871888-4177b980-12b4-11ea-9296-79d5eab47c47.png)

Note again that the last value of the column `mse` is the same as the mean of the column `loss`.

### Output of the console while training:

```
Epoch 1/1
9273/9273 [==============================] - 74s 8ms/step - 
loss: 0.0353 - mae: 0.0870 - mse: 0.0353 - 
val_loss: 0.0015 - val_mae: 0.1338 - val_mse: 0.0517
```

### Bottomline

Checking `loss` in the output from the console, it is displaying the running average of the training loss (green line in the plot for training loss per batch, which is the mean of the blue line). 

However, checking `val_loss` in the output from the console, it is displaying the **loss for the last validation batch** (last value of the blue line in the plot for validation loss per batch). I would expect to obtain the average of the blue line, i.e., the last value of the green line in the same plot.

If I made any mistake or misunderstood anything, please point it out.

**Comment by DuyguA** [comment]
Quick question: Is your data shuffled enough?
Where do the training and validation data come from? Can it be that, you didn't shuffled the dataset before splitting into validation and training?

**Comment by iipr** [comment]
> Quick question: Is your data shuffled enough?

It is not shuffled at all. As I was mentioning it is time dependent and the aim is to eventually implement a recurrent network, so the data needs to be in order.

> Where do the training and validation data come from?

It comes from an HDF5 file which contains the arrays with a certain structure (year -> months -> days). But eventually there is one table per day.

> Can it be that, you didn't shuffled the dataset before splitting into validation and training?

As I say, I don't shuffle the data. The analysis I made suggest that the validation loss displayed during training (and recorded in the history object) is not the average over batches, but the loss of the last validation batch, but I may be wrong in something.

**Comment by DuyguA** [comment]
> As I was mentioning it is time dependent and the aim is to eventually implement a recurrent network, so the data needs to be in order

I didn't quite get this point, do you want to go for `stateful=True` ?

> so the data needs to be in order.

I didn't mean batch shuffling, if you want to make a stateful RNN of course you do not shuffle batches. What I mean is, shuffling your data once before the training, otherwise this is quite possible:

```
data:
class A
class A
class A
..
class B
class B

...
class C
class C
```

then 

```
training: class A, class A, class A, ... class B, class B, class B
validation: class C, class C, class C
```
If you shuffle the dataset then

```
data: class A class B class B class C class A
train: class B class A class C
validation: class A class B class C
```

which is more convenient.

**Comment by iipr** [comment]
>  I didn't quite get this point, do you want to go for stateful=True ?

Yes, I think that I will use a stateful RNN later on.

Regarding the second part, I see your point. I am working on a regression problem, not a classification one: predicting some sensors' measurements (which depend on the time of the day). So by now what I am doing is leaving the last days from each month for validation, using the first ones for training. 

In any case, I don't think this is causing the issue. Checking the numbers I reported above, the `val_loss` displayed is equal to the loss of the last validation batch and not the average over batches:

```
Epoch 1/1
9273/9273 [==============================] - 74s 8ms/step - 
loss: 0.0353 - mae: 0.0870 - mse: 0.0353 - 
val_loss: 0.0015 - val_mae: 0.1338 - val_mse: 0.0517
```
![image](https://user-images.githubusercontent.com/17690149/69871888-4177b980-12b4-11ea-9296-79d5eab47c47.png)

**Comment by iipr** [comment]
This is probably related to https://github.com/keras-team/keras/issues/13641. As discussed in [this comment](https://github.com/keras-team/keras/issues/13641#issuecomment-567513035) of the issue, the problem (bug?) could be [here](https://github.com/keras-team/keras/blob/master/keras/engine/training_generator.py#L420).

**Comment by Young-won** [comment]
I have a similar issue, and the problem could be [here](https://github.com/keras-team/keras/blob/7a39b6c62d43c25472b2c2476bd2a8983ae4f682/keras/engine/training_arrays.py#L418).
I think those lines
```
if i == 0:  # Index 0 == `Loss`
    outs[i] = float(batch_out)
else:
    outs[i] += float(batch_out)
```
shoud be fixed as
```
if i == 0:  # Index 0 == `Loss`
    outs[i] += float(batch_out)
else:
    outs[i] = float(batch_out)
```
